### ISSUE TITLE
NaNs passed as initial point to f, g!

### ISSUE THREAD

**Issue by afbarnard** [JuliaNLSolvers/Optim.jl#646]
Optim.jl no longer works for me as the initial point gets set to an array of NaNs. Here's my situation:

```julia
import Optim
function f(x)
    @show x
    return sum(x .^ 2)
end
function g!(x, g)
    @show x
    @show g
    g[:] = 2 .* x
end
opt = Optim.optimize(f, g!, ones(4), Optim.LBFGS(), Optim.Options(iterations=10, g_tol=1e-6, store_trace=true, show_trace=true))
````

This results in the following output:

```
x = [NaN, NaN, NaN, NaN]
g = [1.0, 1.0, 1.0, 1.0]
x = [NaN, NaN, NaN, NaN]
Iter     Function value   Gradient norm 
     0              NaN              NaN
ERROR: ArgumentError: Value and slope at step length = 0 must be finite.
Stacktrace:
 [1] (::LineSearches.HagerZhang{Float64,Base.RefValue{Bool}})(::LineSearches.#ϕ#2{Optim.ManifoldObjective{NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}},Array{Float64,1},Array{Float64,1},Array{Float64,1}}, ::LineSearches.#ϕdϕ#17{Optim.ManifoldObjective{NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}},Array{Float64,1},Array{Float64,1},Array{Float64,1}}, ::Float64, ::Float64, ::Float64) at /home/barnard/.julia/v0.6/LineSearches/src/hagerzhang.jl:116
 [2] (::LineSearches.HagerZhang{Float64,Base.RefValue{Bool}})(::Optim.ManifoldObjective{NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}}, ::Array{Float64,1}, ::Array{Float64,1}, ::Float64, ::Array{Float64,1}, ::Float64, ::Float64) at /home/barnard/.julia/v0.6/LineSearches/src/hagerzhang.jl:101
 [3] perform_linesearch!(::Optim.LBFGSState{Array{Float64,1},Array{Array{Float64,1},1},Array{Array{Float64,1},1},Float64,Array{Float64,1}}, ::Optim.LBFGS{Void,LineSearches.InitialStatic{Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Optim.##32#34}, ::Optim.ManifoldObjective{NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}}) at /home/barnard/.julia/v0.6/Optim/src/utilities/perform_linesearch.jl:40
 [4] update_state!(::NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}, ::Optim.LBFGSState{Array{Float64,1},Array{Array{Float64,1},1},Array{Array{Float64,1},1},Float64,Array{Float64,1}}, ::Optim.LBFGS{Void,LineSearches.InitialStatic{Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Optim.##32#34}) at /home/barnard/.julia/v0.6/Optim/src/multivariate/solvers/first_order/l_bfgs.jl:198
 [5] optimize(::NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}, ::Array{Float64,1}, ::Optim.LBFGS{Void,LineSearches.InitialStatic{Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Optim.##32#34}, ::Optim.Options{Float64,Void}, ::Optim.LBFGSState{Array{Float64,1},Array{Array{Float64,1},1},Array{Array{Float64,1},1},Float64,Array{Float64,1}}) at /home/barnard/.julia/v0.6/Optim/src/multivariate/optimize/optimize.jl:57
 [6] #optimize#132(::Bool, ::Symbol, ::Function, ::Function, ::Function, ::Array{Float64,1}, ::Optim.LBFGS{Void,LineSearches.InitialStatic{Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Optim.##32#34}, ::Optim.Options{Float64,Void}) at /home/barnard/.julia/v0.6/Optim/src/multivariate/optimize/interface.jl:121
 [7] optimize(::Function, ::Function, ::Array{Float64,1}, ::Optim.LBFGS{Void,LineSearches.InitialStatic{Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Optim.##32#34}, ::Optim.Options{Float64,Void}) at /home/barnard/.julia/v0.6/Optim/src/multivariate/optimize/interface.jl:119
```

I just upgraded everything:

```
Julia Version 0.6.4
Commit 9d11f62bcb (2018-07-09 19:09 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: AMD Athlon(tm) 64 X2 Dual Core Processor 6400+
  WORD_SIZE: 64
           "Fedora release 27 (Twenty Seven)"
  uname: Linux 4.17.3-100.fc27.x86_64 #1 SMP Tue Jun 26 14:19:03 UTC 2018 x86_64 x86_64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Opteron MAX_THREADS=16)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, k8-sse3)
 - Optim                         0.15.2
```

**Comment by afbarnard** [comment]
Seems like the initial `x` gets discarded here(?): https://github.com/JuliaNLSolvers/NLSolversBase.jl/blob/40f448c63d9e625b81bb77d3adf75830bdafaee4/src/objective_types/oncedifferentiable.jl#L163

**Comment by afbarnard** [comment]
Ah, I see that the order of arguments to `g!` was reversed in v0.9.0. Sorry. (My eyes glossed over that difference in the documentation.) I would be nice to use types to enforce the order of arguments to `g!`, but since the only difference is mutability, that would have involve traits or array views or similar future features.

**Comment by anriseth** [comment]
Out of interest, when was the last time you used Optim?

**Comment by pkofod** [comment]
This changed last summer I believe, but good you figured it out :) we actually had a warning for this based on which array changed values when updating the caches the first time :)

**Comment by afbarnard** [comment]
I had been using Optim 0.6.1. So it was a big jump to 0.15.2. I was stuck on Julia 0.5 for various reasons related to my academic work, and, while I had tried to update Optim, there was some problem with 0.7 and the introduction of the package for line searches, so I reverted and stayed with 0.6.1. So, while I have been using Optim since June 2015, I was just using an old version. Why change what was working for me, right? But I finally had the opportunity to put some work into upgrading my code base, so here I am. And things are working again now with all the latest stable versions.
